### PR TITLE
Handle the failure in AssignResGroupOnMaster()

### PR DESCRIPTION
**Pull request by xiong-gang**
As AssignResGroupOnMaster() is called before the transaction is
actually started, so the failure won't cause transaction abort,
we need handle the error to prevent slot leaking.

Signed-off-by: Zhenghua Lyu <zlv@pivotal.io>

### CODE REVIEW

File: src/test/isolation2/expected/resgroup/resgroup_assign_slot_fail.out [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+CREATE
+
+1: SET ROLE role_test;
+SET
+1: BEGIN;
+BEGIN
+2: SET ROLE role_test;
+SET
+SELECT gp_inject_fault('resgroup_assgined_on_master', 'error', 1);
+ERROR:  Failure: could not insert fault injection, entry already exists (gp_inject_fault.c:67)
```
Comment (guofengrichard) [inline src/test/isolation2/expected/resgroup/resgroup_assign_slot_fail.out]: Is this message the expected output for the fault injection?

File: src/backend/access/transam/xact.c [hunk 2]
```diff
@@ -2236,6 +2236,10 @@ StartTransaction(void)
 		SIMPLE_FAULT_INJECTOR(TransactionStartUnderEntryDbSingleton);
 	}
 
+	/* Acquire a resource group slot at the beginning of a transaction */
```
Comment (asimrp) [inline src/backend/access/transam/xact.c]: Can we please add a comment explaining why this needs to be done before `CurrentTransactionState` is initialized?

Comment (xiong-gang) [reply src/backend/access/transam/xact.c]: Thanks, comments added.

File: src/backend/access/transam/xact.c [hunk 3]
```diff
@@ -2237,6 +2237,18 @@ StartTransaction(void)
 	}
 
 	/*
+	 * Acquire a resource group slot.
+	 *
+	 * AssignResGroupOnMaster() might throw error, so call it before touch
+	 * transaction state.
+	 * Slot is successfully acquired when AssignResGroupOnMaster() is returned,
+	 * this slot will be release when transaction is committed or abortted,
```
Comment (danielgustafsson) [inline src/backend/access/transam/xact.c]: s/abortted/aborted/

File: src/backend/utils/resgroup/resgroup.c [hunk 4]
```diff
@@ -1568,7 +1573,7 @@ groupGetMemSpillTotal(const ResGroupCaps *caps)
 static int32
 slotGetMemQuotaExpected(const ResGroupCaps *caps)
 {
-	return groupGetMemQuotaExpected(caps) / caps->concurrency.proposed;
+	return Max(1, groupGetMemQuotaExpected(caps) / caps->concurrency.proposed);
```
Comment (danielgustafsson) [inline src/backend/utils/resgroup/resgroup.c]: While not introduced in this patch, I wonder if we shouldn't move the assertion that the `proposed` member is non-zero to this function? Some codepaths leading here tests for zero, but not all, so testing locally before performing division seems safer.

File: src/backend/utils/misc/faultinjector.c [hunk 5]
```diff
@@ -329,6 +329,8 @@ FaultInjectorIdentifierEnumToString[] = {
 		/* inject fault in FinishPreparedTransaction() after recording the commit prepared record */
 	_("gang_created"),
 		/* inject fault to report ERROR just after creating Gang */
+	_("resgroup_assgined_on_master"),
```
Comment (danielgustafsson) [inline src/backend/utils/misc/faultinjector.c]: s/assgined/assigned/

File: src/backend/utils/resgroup/resgroup.c (hunk 4 above)
Comment (xiong-gang) [reply src/backend/utils/resgroup/resgroup.c]: This `proposed` is from catalog table and it's not supposed to be zero, so yes i will add an assertion here. Thanks

**Comment by cjcjameson** [comment]
Is this the sort of commit that should be backported to 5X_STABLE? Feel free to do so.

Perhaps you're waiting until after 5.0.0 GA; is that the case? If it does go in after the 5.0.0 GA tag goes down, I'm guessing this is a patch-level version bump, not a minor version bump -- can you assess and confirm?

**Comment by xiong-gang** [comment]
Thanks @cjcjameson , i should backport this to 5x_stable

**Comment by xiong-gang** [comment]
Hi @cjcjameson, for commits like this, do we need to file a PR for 5x_stable as well?

**Comment by cjcjameson** [comment]
I don't think so, my recommendation is to do it without a PR.

**Comment by Eulerizeit** [comment]
When we discussed this in the C&S LL it was basically.
1) if everything applies cleanly no PR
2) small fiddling to heavy lifting then use your best judgement

-- Rob

**Comment by xiong-gang** [comment]
Thanks @cjcjameson  @Eulerizeit

**Pull request merged**